File: packages/js-sdk/test-sdk.js
#!/usr/bin/env node

/**
 * Comprehensive SDK Test Suite
 * Tests recent fixes:
 * - Organization parameter (should search all orgs when omitted)
 * - Topic parameter (should default to "tc-ready")
 * - Prompt improvements for translation notes
 */

import { TranslationHelpsClient } from './dist/index.js';

const SERVER_URL = process.env.SERVER_URL || 'http://localhost:3000/api';

const colors = {
  green: '\x1b[32m',
  red: '\x1b[31m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  reset: '\x1b[0m'
};

function log(message, color = colors.reset) {
  console.log(`${color}${message}${colors.reset}`);
}

function logTest(name) {
  console.log(`\n${colors.blue}━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━${colors.reset}`);
  log(`Testing: ${name}`, colors.yellow);
  console.log(`${colors.blue}━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━${colors.reset}`);
}

async function testOrganizationParameter() {
  logTest('Organization Parameter - Multi-Org Search (Default)');
  
  const client = new TranslationHelpsClient({ serverUrl: SERVER_URL });
  await client.connect();
  
  try {
    log('Testing list_resources_for_language WITHOUT organization param...', colors.blue);
    const result = await client.listResourcesForLanguage({
      language: 'en',
      limit: 100
    });
    
    const orgCounts = {};
    for (const [subject, resources] of Object.entries(result.resourcesBySubject)) {
      for (const resource of resources) {
        const org = resource.organization || 'unknown';
        orgCounts[org] = (orgCounts[org] || 0) + 1;
      }
    }
    
    log(`✓ Found resources from ${Object.keys(orgCounts).length} organizations:`, colors.green);
    for (const [org, count] of Object.entries(orgCounts)) {
      console.log(`  - ${org}: ${count} resources`);
    }
    
    if (Object.keys(orgCounts).length > 1) {
      log('✓ PASS: Multi-organization search working correctly!', colors.green);
    } else {
      log('⚠ WARNING: Only found 1 organization. This might be correct if only one org has "en" resources.', colors.yellow);
    }
    
  } catch (error) {
    log(`✗ FAIL: ${error.message}`, colors.red);
    throw error;
  }
}

async function testOrganizationParameterSpecific() {
  logTest('Organization Parameter - Single Org Filter');
  
  const client = new TranslationHelpsClient({ serverUrl: SERVER_URL });
  await client.connect();
  
  try {
    log('Testing list_resources_for_language WITH organization="unfoldingWord"...', colors.blue);
    const result = await client.listResourcesForLanguage({
      language: 'en',
      organization: 'unfoldingWord',
      limit: 100
    });
    
    const orgCounts = {};
    for (const [subject, resources] of Object.entries(result.resourcesBySubject)) {
      for (const resource of resources) {
        const org = resource.organization || 'unknown';
        orgCounts[org] = (orgCounts[org] || 0) + 1;
      }
    }
    
    log(`✓ Found resources from ${Object.keys(orgCounts).length} organization(s):`, colors.green);
    for (const [org, count] of Object.entries(orgCounts)) {
      console.log(`  - ${org}: ${count} resources`);
    }
    
    if (Object.keys(orgCounts).length === 1 && orgCounts['unfoldingWord']) {
      log('✓ PASS: Single organization filter working correctly!', colors.green);
    } else {
      log('✗ FAIL: Expected only "unfoldingWord" resources', colors.red);
      throw new Error('Organization filter not working correctly');
    }
    
  } catch (error) {
    log(`✗ FAIL: ${error.message}`, colors.red);
    throw error;
  }
}

async function testTopicParameter() {
  logTest('Topic Parameter - Default "tc-ready"');
  
  const client = new TranslationHelpsClient({ serverUrl: SERVER_URL });
  await client.connect();
  
  try {
    log('Testing list_resources_for_language WITHOUT topic param (should default to "tc-ready")...', colors.blue);
    const result = await client.listResourcesForLanguage({
      language: 'en',
      limit: 50
    });
    
    const topicCounts = {};
    for (const [subject, resources] of Object.entries(result.resourcesBySubject)) {
      for (const resource of resources) {
        const topics = resource.metadata?.topic || ['no-topic'];
        for (const topic of topics) {
          topicCounts[topic] = (topicCounts[topic] || 0) + 1;
        }
      }
    }
    
    log(`✓ Found resources with topics:`, colors.green);
    for (const [topic, count] of Object.entries(topicCounts)) {
      console.log(`  - ${topic}: ${count} resources`);
    }
    
    if (topicCounts['tc-ready'] > 0) {
      log('✓ PASS: Topic filter defaulting to "tc-ready" correctly!', colors.green);
    } else {
      log('⚠ WARNING: No "tc-ready" resources found. This might be correct for some languages.', colors.yellow);
    }
    
  } catch (error) {
    log(`✗ FAIL: ${error.message}`, colors.red);
    throw error;
  }
}

async function testFetchScripture() {
  logTest('Fetch Scripture - Organization Parameter');
  
  const client = new TranslationHelpsClient({ serverUrl: SERVER_URL });
  await client.connect();
  
  try {
    log('Testing fetchScripture WITHOUT organization (should search all orgs)...', colors.blue);
    const scripture = await client.fetchScripture({
      reference: 'John 3:16',
      language: 'en',
      format: 'text'
    });
    
    if (scripture && scripture.length > 0) {
      log(`✓ Successfully fetched scripture: "${scripture.substring(0, 50)}..."`, colors.green);
      log('✓ PASS: Scripture fetch working without organization!', colors.green);
    } else {
      throw new Error('No scripture text returned');
    }
    
  } catch (error) {
    log(`✗ FAIL: ${error.message}`, colors.red);
    throw error;
  }
}

async function testFetchTranslationNotes() {
  logTest('Fetch Translation Notes - Organization Parameter');
  
  const client = new TranslationHelpsClient({ serverUrl: SERVER_URL });
  await client.connect();
  
  try {
    log('Testing fetchTranslationNotes WITHOUT organization...', colors.blue);
    const notes = await client.fetchTranslationNotes({
      reference: 'Titus 1:1',
      language: 'en'
    });
    
    if (notes) {
      log(`✓ Successfully fetched translation notes`, colors.green);
      console.log(`  - Verse notes: ${notes.verseNotes?.length || 0}`);
      console.log(`  - Context notes: ${notes.contextNotes?.length || 0}`);
      
      if (notes.verseNotes && notes.verseNotes.length > 0) {
        log(`\nFirst note preview:`, colors.blue);
        const firstNote = notes.verseNotes[0];
        console.log(`  Quote: ${firstNote.Quote}`);
        console.log(`  Note: ${firstNote.Note?.substring(0, 100)}...`);
      }
      
      log('✓ PASS: Translation notes fetch working without organization!', colors.green);
    } else {
      throw new Error('No notes returned');
    }
    
  } catch (error) {
    log(`✗ FAIL: ${error.message}`, colors.red);
    throw error;
  }
}

async function testFetchTranslationQuestions() {
  logTest('Fetch Translation Questions - Organization Parameter');
  
  const client = new TranslationHelpsClient({ serverUrl: SERVER_URL });
  await client.connect();
  
  try {
    log('Testing fetchTranslationQuestions WITHOUT organization...', colors.blue);
    const questions = await client.fetchTranslationQuestions({
      reference: 'Genesis 1:1',
      language: 'en'
    });
    
    if (questions && questions.length > 0) {
      log(`✓ Successfully fetched ${questions.length} translation questions`, colors.green);
      log(`\nFirst question preview:`, colors.blue);
      console.log(`  Q: ${questions[0].Question}`);
      console.log(`  A: ${questions[0].Answer?.substring(0, 100)}...`);
      log('✓ PASS: Translation questions fetch working without organization!', colors.green);
    } else {
      throw new Error('No questions returned');
    }
    
  } catch (error) {
    log(`✗ FAIL: ${error.message}`, colors.red);
    throw error;
  }
}

async function testLanguageVariantFallback() {
  logTest('Language Variant Fallback (es-419 → es)');
  
  const client = new TranslationHelpsClient({ serverUrl: SERVER_URL });
  await client.connect();
  
  try {
    log('Testing list_resources_for_language with es-419...', colors.blue);
    const result = await client.listResourcesForLanguage({
      language: 'es-419',
      limit: 50
    });
    
    if (result.totalResources > 0) {
      log(`✓ Successfully found ${result.totalResources} resources`, colors.green);
      log(`✓ Language used: ${result.language || 'es-419'}`, colors.green);
      
      if (result.detectedVariant) {
        log(`✓ Variant fallback detected: ${JSON.stringify(result.detectedVariant)}`, colors.green);
      }
      
      log('✓ PASS: Language variant fallback working!', colors.green);
    } else {
      log('⚠ WARNING: No resources found for es-419', colors.yellow);
    }
    
  } catch (error) {
    log(`✗ FAIL: ${error.message}`, colors.red);
    throw error;
  }
}

async function runAllTests() {
  console.log(`\n${colors.blue}╔═══════════════════════════════════════════════════════╗${colors.reset}`);
  console.log(`${colors.blue}║     Translation Helps SDK - Comprehensive Tests      ║${colors.reset}`);
  console.log(`${colors.blue}╚═══════════════════════════════════════════════════════╝${colors.reset}`);
  console.log(`\nServer URL: ${SERVER_URL}\n`);
  
  const tests = [
    { name: 'Multi-Org Search (Default)', fn: testOrganizationParameter },
    { name: 'Single Org Filter', fn: testOrganizationParameterSpecific },
    { name: 'Topic Default', fn: testTopicParameter },
    { name: 'Fetch Scripture', fn: testFetchScripture },
    { name: 'Fetch Translation Notes', fn: testFetchTranslationNotes },
    { name: 'Fetch Translation Questions', fn: testFetchTranslationQuestions },
    { name: 'Language Variant Fallback', fn: testLanguageVariantFallback }
  ];
  
  let passed = 0;
  let failed = 0;
  
  for (const test of tests) {
    try {
      await test.fn();
      passed++;
    } catch (error) {
      failed++;
      log(`\n✗ Test "${test.name}" failed: ${error.message}`, colors.red);
    }
  }
  
  console.log(`\n${colors.blue}━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━${colors.reset}`);
  log(`\n📊 Test Results:`, colors.yellow);
  log(`  ✓ Passed: ${passed}/${tests.length}`, passed === tests.length ? colors.green : colors.yellow);
  if (failed > 0) {
    log(`  ✗ Failed: ${failed}/${tests.length}`, colors.red);
  }
  console.log(`\n${colors.blue}━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━${colors.reset}\n`);
  
  if (failed > 0) {
    process.exit(1);
  }
}

// Run tests
runAllTests().catch(error => {
  log(`\n✗ Fatal error: ${error.message}`, colors.red);
  console.error(error);
  process.exit(1);
});

File: packages/python-sdk/test_sdk.py
#!/usr/bin/env python3
"""
Comprehensive SDK Test Suite
Tests recent fixes:
- list_resources_for_language searches all orgs (no org filter param)
- Topic parameter (should default to "tc-ready")
- Prompt improvements for translation notes
"""

import asyncio
import os
import sys
from typing import Any, Dict

# Add the package to the path
sys.path.insert(0, os.path.dirname(__file__))

from translation_helps import TranslationHelpsClient

SERVER_URL = os.getenv('SERVER_URL', 'http://localhost:3000/api')

class Colors:
    GREEN = '\033[32m'
    RED = '\033[31m'
    YELLOW = '\033[33m'
    BLUE = '\033[34m'
    RESET = '\033[0m'

def log(message: str, color: str = Colors.RESET):
    print(f"{color}{message}{Colors.RESET}")

def log_test(name: str):
    print(f"\n{Colors.BLUE}========================================{Colors.RESET}")
    log(f"Testing: {name}", Colors.YELLOW)
    print(f"{Colors.BLUE}========================================{Colors.RESET}")

async def test_organization_parameter():
    log_test('Organization Parameter - Multi-Org Search (Default)')
    
    client = TranslationHelpsClient({"serverUrl": SERVER_URL})
    await client.connect()
    
    try:
        log('Testing list_resources_for_language WITHOUT organization param...', Colors.BLUE)
        result = await client.list_resources_for_language({
            'language': 'en',
            'limit': 100
        })
        
        org_counts = {}
        for subject, resources in result.get('resourcesBySubject', {}).items():
            for resource in resources:
                org = resource.get('organization', 'unknown')
                org_counts[org] = org_counts.get(org, 0) + 1
        
        log(f"[OK] Found resources from {len(org_counts)} organizations:", Colors.GREEN)
        for org, count in org_counts.items():
            print(f"  - {org}: {count} resources")
        
        if len(org_counts) > 1:
            log('[OK] PASS: Multi-organization search working correctly!', Colors.GREEN)
        else:
            log('[!] WARNING: Only found 1 organization. This might be correct if only one org has "en" resources.', Colors.YELLOW)
        
    except Exception as error:
        log(f'[X] FAIL: {str(error)}', Colors.RED)
        raise
    finally:
        await client.close()

async def test_topic_parameter():
    log_test('Topic Parameter - Default "tc-ready"')
    
    client = TranslationHelpsClient({"serverUrl": SERVER_URL})
    await client.connect()
    
    try:
        log('Testing list_resources_for_language WITHOUT topic param (should default to "tc-ready")...', Colors.BLUE)
        result = await client.list_resources_for_language({
            'language': 'en',
            'limit': 50
        })
        
        topic_counts = {}
        for subject, resources in result.get('resourcesBySubject', {}).items():
            for resource in resources:
                topics = resource.get('metadata', {}).get('topic', ['no-topic'])
                for topic in topics:
                    topic_counts[topic] = topic_counts.get(topic, 0) + 1
        
        log("[OK] Found resources with topics:", Colors.GREEN)
        for topic, count in topic_counts.items():
            print(f"  - {topic}: {count} resources")
        
        if topic_counts.get('tc-ready', 0) > 0:
            log('[OK] PASS: Topic filter defaulting to "tc-ready" correctly!', Colors.GREEN)
        else:
            log('[!] WARNING: No "tc-ready" resources found. This might be correct for some languages.', Colors.YELLOW)
        
    except Exception as error:
        log(f'[X] FAIL: {str(error)}', Colors.RED)
        raise
    finally:
        await client.close()

async def test_fetch_scripture():
    log_test('Fetch Scripture - Organization Parameter')
    
    client = TranslationHelpsClient({"serverUrl": SERVER_URL})
    await client.connect()
    
    try:
        log('Testing fetch_scripture WITHOUT organization (should search all orgs)...', Colors.BLUE)
        scripture = await client.fetch_scripture({
            'reference': 'John 3:16',
            'language': 'en',
            'format': 'text'
        })
        
        if scripture and len(scripture) > 0:
            log(f'[OK] Successfully fetched scripture: "{scripture[:50]}..."', Colors.GREEN)
            log('[OK] PASS: Scripture fetch succeeded', Colors.GREEN)
        else:
            raise Exception('No scripture text returned')
        
    except Exception as error:
        log(f'[X] FAIL: {str(error)}', Colors.RED)
        raise
    finally:
        await client.close()

async def test_fetch_translation_notes():
    log_test('Fetch Translation Notes')
    
    client = TranslationHelpsClient({"serverUrl": SERVER_URL})
    await client.connect()
    
    try:
        log('Testing fetch_translation_notes...', Colors.BLUE)
        notes = await client.fetch_translation_notes({
            'reference': 'Titus 1:1',
            'language': 'en'
        })
        
        if notes:
            log("[OK] Successfully fetched translation notes", Colors.GREEN)
            print(f"  - Verse notes: {len(notes.get('verseNotes', []))}")
            print(f"  - Context notes: {len(notes.get('contextNotes', []))}")
            
            verse_notes = notes.get('verseNotes', [])
            if verse_notes and len(verse_notes) > 0:
                log("\nFirst note preview:", Colors.BLUE)
                first_note = verse_notes[0]
                try:
                    print(f"  Quote: {first_note.get('Quote')}")
                    note_text = first_note.get('Note', '')
                    print(f"  Note: {note_text[:100]}...")
                except UnicodeEncodeError:
                    print(f"  Quote: [Greek text - encoding issue]")
                    print(f"  Note: {first_note.get('Note', '')[:100].encode('ascii', 'replace').decode()}...")
            
            log('[OK] PASS: Translation notes fetch succeeded', Colors.GREEN)
        else:
            raise Exception('No notes returned')
        
    except Exception as error:
        log(f'[X] FAIL: {str(error)}', Colors.RED)
        raise
    finally:
        await client.close()

async def test_fetch_translation_questions():
    log_test('Fetch Translation Questions')
    
    client = TranslationHelpsClient({"serverUrl": SERVER_URL})
    await client.connect()
    
    try:
        log('Testing fetch_translation_questions...', Colors.BLUE)
        questions = await client.fetch_translation_questions({
            'reference': 'Genesis 1:1',
            'language': 'en'
        })
        
        if questions and len(questions) > 0:
            log(f'[OK] Successfully fetched {len(questions)} translation questions', Colors.GREEN)
            log("\nFirst question preview:", Colors.BLUE)
            try:
                print(f"  Q: {questions[0].get('Question')}")
                answer = questions[0].get('Answer', '')
                print(f"  A: {answer[:100]}...")
            except (UnicodeEncodeError, IndexError):
                print(f"  Q: [Question text - encoding/index issue]")
                print(f"  A: [Answer text - encoding/index issue]")
            log('[OK] PASS: Translation questions fetch succeeded', Colors.GREEN)
        else:
            raise Exception('No questions returned')
        
    except Exception as error:
        log(f'[X] FAIL: {str(error)}', Colors.RED)
        raise
    finally:
        await client.close()

async def test_language_variant_fallback():
    log_test('Language Variant Fallback (es-419 to es)')
    
    client = TranslationHelpsClient({"serverUrl": SERVER_URL})
    await client.connect()
    
    try:
        log('Testing list_resources_for_language with es-419...', Colors.BLUE)
        result = await client.list_resources_for_language({
            'language': 'es-419',
            'limit': 50
        })
        
        total = result.get('totalResources', 0)
        if total > 0:
            log(f'[OK] Successfully found {total} resources', Colors.GREEN)
            log(f'[OK] Language used: {result.get("language", "es-419")}', Colors.GREEN)
            
            if result.get('detectedVariant'):
                log(f'[OK] Variant fallback detected: {result.get("detectedVariant")}', Colors.GREEN)
            
            log('[OK] PASS: Language variant fallback working!', Colors.GREEN)
        else:
            log('[!] WARNING: No resources found for es-419', Colors.YELLOW)
        
    except Exception as error:
        log(f'[X] FAIL: {str(error)}', Colors.RED)
        raise
    finally:
        await client.close()

async def run_all_tests():
    print(f"\n{Colors.BLUE}==============================================={Colors.RESET}")
    print(f"{Colors.BLUE}  Translation Helps SDK - Comprehensive Tests  {Colors.RESET}")
    print(f"{Colors.BLUE}==============================================={Colors.RESET}")
    print(f"\nServer URL: {SERVER_URL}\n")
    
    tests = [
        {'name': 'Multi-Org Discovery', 'fn': test_multi_organization_discovery},
        {'name': 'Topic Default', 'fn': test_topic_parameter},
        {'name': 'Fetch Scripture', 'fn': test_fetch_scripture},
        {'name': 'Fetch Translation Notes', 'fn': test_fetch_translation_notes},
        {'name': 'Fetch Translation Questions', 'fn': test_fetch_translation_questions},
        {'name': 'Language Variant Fallback', 'fn': test_language_variant_fallback}
    ]
    
    passed = 0
    failed = 0
    
    for test in tests:
        try:
            await test['fn']()
            passed += 1
        except Exception as error:
            failed += 1
            log(f'\n[FAIL] Test "{test["name"]}" failed: {str(error)}', Colors.RED)
    
    print(f"\n{Colors.BLUE}========================================{Colors.RESET}")
    log("\nTest Results:", Colors.YELLOW)
    color = Colors.GREEN if passed == len(tests) else Colors.YELLOW
    log(f"  Passed: {passed}/{len(tests)}", color)
    if failed > 0:
        log(f"  Failed: {failed}/{len(tests)}", Colors.RED)
    print(f"\n{Colors.BLUE}========================================{Colors.RESET}\n")
    
    if failed > 0:
        sys.exit(1)

if __name__ == '__main__':
    try:
        asyncio.run(run_all_tests())
    except KeyboardInterrupt:
        log('\n\n[INTERRUPTED] Tests interrupted by user', Colors.RED)
        sys.exit(1)
    except Exception as error:
        log(f'\n[ERROR] Fatal error: {str(error)}', Colors.RED)
        import traceback
        traceback.print_exc()
        sys.exit(1)
